src/passes/FloatTrap.cpp

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright 2017 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "passes/FloatTrap.h"
+
+#include "asm_v_wasm.h"
+#include "mixed_arena.h"
+#include "wasm-builder.h"
+#include "wasm-printing.h"
+#include "wasm-type.h"
+#include "asmjs/shared-constants.h"
+
+namespace wasm {
+
+Name I64S_REM("i64s-rem"),
+     I64U_REM("i64u-rem"),
+     I64S_DIV("i64s-div"),
+     I64U_DIV("i64u-div");
+
+FloatTrap::FloatTrap(Mode mode)
+    : mode(mode),
+      didAddF64ToI64JSImport(false) {
+  name = "binaryen-trap-mode";
+}
+
+void FloatTrap::visitUnary(Unary* curr) {
+  replaceCurrent(makeTrappingUnary(curr));
+}
+
+void FloatTrap::visitBinary(Binary* curr) {
+  replaceCurrent(makeTrappingBinary(curr));
+}
+
+void FloatTrap::visitModule(Module* curr) {
+  for (const auto& pair : generatedFunctions) {
+    curr->addFunction(pair.second);
+  }
+  generatedFunctions.clear();
+}
+
+Expression* FloatTrap::makeTrappingBinary(Binary* curr) {
+  Name name = getBinaryFuncName(curr);
+  if (mode == Mode::Allow || !name.is()) {
+    return curr;
+  }
+
+  // the wasm operation might trap if done over 0, so generate a safe call
+  WasmType type = curr->type;
+  MixedArena& allocator(getModule()->allocator);
+  auto *call = allocator.alloc<Call>();
+  call->target = name;
+  call->operands.push_back(curr->left);
+  call->operands.push_back(curr->right);
+  call->type = type;
+  ensureBinaryFunc(curr);
+  return call;
+}
+
+Expression* FloatTrap::makeTrappingUnary(Unary* curr) {
+  Name name = getUnaryFuncName(curr);
+  if (mode == Mode::Allow || !name.is()) {
```

Comment:
Would make sense the call to `getUnaryFunctionName` conditional on the mode being something other than allow? Then we wouldn't run the switch statement unless we knew we actually had to do something.

Come to think of it, why do we even have the allow mode? Doesn't that make the whole pass a no-op?

Reply:
Allow does make the whole pass a no-op. It seems cleaner than conditionally adding the pass if the flag is set.
Could probably do something more clever by not walking the module if Allow, by overriding one of the Walker methods. Would still introduce a sequence point by being another non-parallel pass.